stablehlo/dialect/StablehloOps.cpp

```diff
@@ -4660,49 +4661,65 @@ LogicalResult ReplicaIdOp::inferReturnTypes(
-           << branchName << " must have 0 arguments, but found "
-           << region.getNumArguments();
-
-  TypeRange branchReturnTypes =
-      region.front().getTerminator()->getOperandTypes();
-  if (branchReturnTypes != op->getResultTypes())
-    return op->emitOpError()
-           << branchName << " returned types (" << branchReturnTypes
-           << ") do not match op result types (" << op->getResultTypes() << ")";
+// TODO: remove after https://github.com/openxla/stablehlo/pull/145 landed.
+bool isCompatibleForTypeRange(TypeRange typeRange1, TypeRange typeRange2) {
+  if (typeRange1.size() != typeRange2.size()) return false;
+  for (auto it : llvm::zip(typeRange1, typeRange2)) {
+    if (!mlir::hlo::isCompatibleForHloTypeInference(std::get<0>(it),
+                                                    std::get<1>(it)))
+      return false;
+  }
+  return true;
+}
 
+static LogicalResult inferConditionalReturnTypeComponents(
+    RegionRange branches, Optional<Location> location,
+    SmallVectorImpl<ShapedTypeComponents>& inferredReturnShapes) {
+  if (branches.empty())
+    return emitOptionalError(location, "expect as lease one branch");
+
+  ValueTypeRange<OperandRange> branch0ResultTypes =
+      branches[0]->front().getTerminator()->getOperandTypes();
+  for (unsigned i = 0; i < branches.size(); ++i) {
+    Twine branchName = "branch " + Twine(i);
+    Region* region = branches[i];
+    if (region->getNumArguments() != 0)
+      return emitOptionalError(location, branchName,
+                               " must have 0 arguments, but found ",
+                               region->getNumArguments());
+
+    auto branchResultTypes = region->front().getTerminator()->getOperandTypes();
+    if (!isCompatibleForTypeRange(branch0ResultTypes, branchResultTypes))
+      return emitOptionalError(location, "branch 0 and ", branchName,
+                               " have mismatched return types: ",
+                               branch0ResultTypes, " vs ", branchResultTypes);
+  }
+  for (auto resultType : branch0ResultTypes)
+    inferredReturnShapes.emplace_back(resultType.cast<ShapedType>());
   return success();
 }
 
-LogicalResult IfOp::verify() {
-  if (failed(verifyConditionalBranch(*this, true_branch(),
-                                     /*branchName=*/"true_branch"))) {
-    return failure();
-  }
-
-  if (failed(verifyConditionalBranch(*this, false_branch(),
-                                     /*branchName=*/"false_branch"))) {
-    return failure();
-  }
-  return success();
+LogicalResult IfOp::inferReturnTypeComponents(
+    MLIRContext*, Optional<Location> location, ValueShapeRange operands,
+    DictionaryAttr attributes, RegionRange regions,
+    SmallVectorImpl<ShapedTypeComponents>& inferredReturnShapes) {
+  IfOp::Adaptor adaptor(operands, attributes, regions);
+  return inferConditionalReturnTypeComponents(
+      ArrayRef<Region*>{&adaptor.true_branch(), &adaptor.false_branch()},
```

Comment:
This can be simplified to `adaptor.getRegions()`.

Reply:
Yes I change this for both IfOp and CaseOp so that they are aligned.